pkg/suggestion/v1beta1/optuna/service.py

```diff
@@ -0,0 +1,178 @@
+                    kwargs["restart_strategy"] = v
+                elif k == "sigma":
+                    kwargs["sigma0"] = float(v)
+                elif k == "random_state":
+                    kwargs["seed"] = int(v)
+                else:
+                    raise ValueError("Unknown name for {}: {}".format(name, k))
+
+            sampler = optuna.samplers.CmaEsSampler(**kwargs)
+
+        elif name == "random":
+            kwargs = {}
+            for k, v in settings.items():
+                if k == "random_state":
+                    kwargs["seed"] = int(v)
+                else:
+                    raise ValueError("Unknown name for {}: {}".format(name, k))
+
+            sampler = optuna.samplers.RandomSampler(**kwargs)
+
+        else:
+            raise ValueError("Unknown algorithm name: {}".format(name))
+
+        return sampler
+
+    def _ask(self, request_number):
+        list_of_assignments = []
+        for _ in range(request_number):
+            optuna_trial = self.study.ask(fixed_distributions=self._get_optuna_search_space())
+
+            assignments = [Assignment(k,v) for k,v in optuna_trial.params.items()]
+            list_of_assignments.append(assignments)
+
+            assignments_key = self._get_assignments_key(assignments)
+            self.assignments_to_optuna_number[assignments_key].append(optuna_trial.number)
+
+        return list_of_assignments
+
+    def _tell(self, trials):
+        for trial in trials:
+            if trial.name not in self.recorded_trial_names:
+                self.recorded_trial_names.add(trial.name)
+
+                value = float(trial.target_metric.value)
+                assignments_key = self._get_assignments_key(trial.assignments)
+                optuna_trial_numbers = self.assignments_to_optuna_number[assignments_key]
+
+                if len(optuna_trial_numbers) != 0:
+                    # The trial has been suggested by the Optuna study.
+                    # The objective value is reported using study.tell() with the corresponding trial number.
+                    trial_number = optuna_trial_numbers.pop(0)
+                    self.study.tell(trial_number, value)
+                else:
+                    # The trial has not been suggested by the Optuna study.
+                    # A new trial object is created and reported using study.add_trial() with the assignments and the search space.
+                    optuna_trial = optuna.create_trial(
+                        params={a.name:self._get_casted_assignment_value(a) for a in trial.assignments},
+                        distributions=self._get_optuna_search_space(),
+                        value=value,
+                    )
+                    self.study.add_trial(optuna_trial)
+
+    def _get_assignments_key(self, assignments):
+        assignments = sorted(assignments, key=lambda a: a.name)
+        assignments_str = [str(a) for a in assignments]
+        return ",".join(assignments_str)
```

Comment:
[Question] I have two questions.

1. Does the string representation of Assigment object holds both of parameter names and parameter values?
1. You defined `assignments_to_optuna_number` as `defaultdict(list)`. I guess the reason why you use list is for duplicated hyperparameters, right?

Reply:
1. Yes; this should contain both the assignment's name and value, using the string representation of the Assignment class.
https://github.com/kubeflow/katib/blob/abbc9c9d34ef0b4ef6b165bb7dc3d6b91ccb5a1d/pkg/suggestion/v1beta1/internal/trial.py#L88-L89 

2. As you mention, this is to handle duplications of assignments.

Reply:
Great! Sounds like it works 👍 

This is a very nit-picking comment but I'd say the following code is more clear to ensure that contains both the parameter name and the value (and a bit memory efficient).

```python
assignments_str = [f"{a.name}:{a.value}" for a in assignments]
```

Reply:
Agree. I changed the line.
https://github.com/kubeflow/katib/pull/1613/commits/b2085a6340cc8e1ef57c70a8956facb296972e6b